File: clients/windows/Mailcal/Services/SignatureImage.cs
// Turning a picked file into the `data:` URI a signature embeds, and saying why it can't, when it
// can't. The pure half lives here so the cap and the media-type refusal are testable without a file
// picker; the WinUI half (Dialogs/SettingsDialog.SignatureEditor.cs) reads the bytes and shows the
// message.
//
// A signature stores its images inline, which is right for the library (one self-contained file) and
// what the core rewrites to a `cid:` part on send, because Outlook's reader blocks `data:` images
// (docs/signatures.md).

using System.Globalization;

namespace Allodia.Mailcal.Services;

/// <summary>The outcome of picking an image for a signature.</summary>
internal abstract record SignatureImageOutcome
{
    /// <summary>A <c>data:image/…;base64,…</c> URI, ready to insert at the caret.</summary>
    /// <param name="Value">The URI.</param>
    /// <param name="AltText">The alt text to give the image (the file's base name).</param>
    internal sealed record DataUrl(string Value, string AltText) : SignatureImageOutcome;

    /// <summary>The file is over the per-image cap; carries the cap so the message can name it.</summary>
    /// <param name="LimitBytes">The cap, in bytes.</param>
    internal sealed record TooLarge(int LimitBytes) : SignatureImageOutcome;

    /// <summary>The file could not be read, or is not an image.</summary>
    internal sealed record Failed : SignatureImageOutcome;
}

/// <summary>The size cap and media-type rule an embedded signature image must pass.</summary>
internal static class SignatureImage
{
    /// <summary>
    /// The cap on an embedded signature image, in bytes. A signature rides in **every** message the
    /// account sends, so a 5 MB logo is 5 MB per mail, and base64 adds a third on top. 512 KB is
    /// generous for a logo and small enough that nobody notices it on the wire. Enforced where the
    /// file is picked, so the user is told; the core does not police it.
    /// </summary>
    internal const int LimitBytes = 512 * 1024;

    /// <summary>
    /// The outcome for <paramref name="bytes"/> of <paramref name="mediaType"/>. The size check is
    /// separate from the read failure so the user is told WHICH problem it is. Anything that is not
    /// an <c>image/*</c> is refused here rather than embedded: the editor would drop it anyway (it
    /// only accepts <c>data:image/</c>), and the picker is where the user can still be told.
    /// </summary>
    internal static SignatureImageOutcome From(byte[] bytes, string? mediaType, string altText)
    {
        if (bytes.Length > LimitBytes)
        {
            return new SignatureImageOutcome.TooLarge(LimitBytes);
        }
        if (bytes.Length == 0
            || mediaType is null
            || !mediaType.StartsWith("image/", StringComparison.OrdinalIgnoreCase))
        {
            return new SignatureImageOutcome.Failed();
        }
        return new SignatureImageOutcome.DataUrl(
            $"data:{mediaType.ToLowerInvariant()};base64,{Convert.ToBase64String(bytes)}",
            altText);
    }

    /// <summary>The cap as the message shows it ("512 KB"), in the current UI culture, the same
    /// decimal-KB convention DiagnosticsLog.FormatBytes and the Android/Apple limits use.</summary>
    internal static string FormatLimit(int bytes) =>
        $"{(bytes / 1000.0).ToString("0.#", CultureInfo.CurrentCulture)} KB";
}

File: clients/linux/src/ui/signature_image.rs
//! The size cap and media-type rule an image embedded in a signature must pass.
//!
//! A signature stores its images inline as `data:` URIs: right for the library (one
//! self-contained file) and what the core rewrites to a `cid:` part on send
//! ([`docs/signatures.md`](../../../../docs/signatures.md)). The pure half lives here so both
//! rules are testable without a file picker; the GTK half reads the bytes and shows the message.

use gtk::glib;

/// The cap on one embedded signature image, in bytes.
///
/// A signature rides in **every** message the account sends, so a 5 MB logo is 5 MB per mail,
/// and base64 adds a third on top. 512 KB is generous for a logo and small enough that nobody
/// notices it on the wire. Enforced where the file is picked, so the user is told; the core does
/// not police it.
pub(super) const LIMIT_BYTES: u64 = 512 * 1024;

/// What picking a file for a signature yielded.
#[derive(Debug, PartialEq, Eq)]
pub(super) enum SignatureImage {
    /// A `data:image/…;base64,…` URI ready to insert at the caret, and the alt text to give it.
    DataUrl { value: String, alt_text: String },
    /// The file is over [`LIMIT_BYTES`].
    TooLarge,
    /// The file could not be read, or is not an image.
    Failed,
}

/// The outcome for `bytes` of `media_type`.
///
/// The size check is separate from the read failure so the user is told *which* problem it is.
/// Anything that is not an `image/*` is refused here rather than embedded: the editor drops it
/// anyway (it accepts only `data:image/`), and the picker is the last place the user can be told.
pub(super) fn signature_image(
    bytes: &[u8],
    media_type: Option<&str>,
    alt_text: &str,
) -> SignatureImage {
    if bytes.len() as u64 > LIMIT_BYTES {
        return SignatureImage::TooLarge;
    }
    let Some(media_type) =
        media_type.filter(|kind| kind.to_ascii_lowercase().starts_with("image/"))
    else {
        return SignatureImage::Failed;
    };
    if bytes.is_empty() {
        return SignatureImage::Failed;
    }
    SignatureImage::DataUrl {
        value: format!(
            "data:{};base64,{}",
            media_type.to_ascii_lowercase(),
            glib::base64_encode(bytes)
        ),
        alt_text: alt_text.to_owned(),
    }
}

/// The cap as the message names it, in the same words this client already gives a file size
/// (the diagnostics log's size row) rather than a second convention.
pub(super) fn format_limit() -> String {
    glib::format_size(LIMIT_BYTES).into()
}

#[cfg(test)]
mod tests {
    use super::{LIMIT_BYTES, SignatureImage, format_limit, signature_image};

    #[test]
    fn an_image_within_the_cap_becomes_a_self_contained_data_uri() {
        let outcome = signature_image(&[0x01, 0x02, 0x03], Some("image/png"), "logo");

        assert_eq!(
            outcome,
            SignatureImage::DataUrl {
                value: "data:image/png;base64,AQID".to_owned(),
                alt_text: "logo".to_owned(),
            }
        );
    }

    #[test]
    fn the_cap_and_the_media_type_refusal_are_told_apart() {
        let oversized = vec![0u8; usize::try_from(LIMIT_BYTES).expect("cap fits a usize") + 1];

        // Two different messages for two different problems: "too large" names a cap the user can
        // meet, "couldn't be read" does not.
        assert_eq!(
            signature_image(&oversized, Some("image/png"), "logo"),
            SignatureImage::TooLarge
        );
        assert_eq!(
            signature_image(&[0x01], Some("application/pdf"), "doc"),
            SignatureImage::Failed
        );
        assert_eq!(
            signature_image(&[0x01], None, "unknown"),
            SignatureImage::Failed
        );
        assert_eq!(
            signature_image(&[], Some("image/png"), ""),
            SignatureImage::Failed
        );
    }

    #[test]
    fn a_media_type_the_server_shouted_still_matches() {
        // `content_type_get_mime_type` answers from the shared MIME database, which is lower case
        //; but the check must not be the thing that decides whether a logo embeds.
        assert!(matches!(
            signature_image(&[0x01], Some("IMAGE/PNG"), "logo"),
            SignatureImage::DataUrl { ref value, .. } if value.starts_with("data:image/png;base64,")
        ));
    }

    #[test]
    fn the_cap_is_named_in_the_message() {
        let limit = format_limit();

        assert!(!limit.is_empty());
        assert!(
            limit.contains(char::is_numeric),
            "the message has to name a size the user can compare against: {limit}"
        );
    }
}
